sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/MapValuesShouldBeAccessedSafelyCheck.kt

```diff
@@ -0,0 +1,65 @@
+/*
+ * SonarSource Kotlin
+ * Copyright (C) 2018-2023 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonarsource.kotlin.checks
+
+import org.jetbrains.kotlin.codegen.kotlinType
+import org.jetbrains.kotlin.js.descriptorUtils.getKotlinTypeFqName
+import org.jetbrains.kotlin.lexer.KtTokens
+import org.jetbrains.kotlin.psi.*
+import org.jetbrains.kotlin.psi.psiUtil.collectDescendantsOfType
+import org.jetbrains.kotlin.psi.psiUtil.getNextSiblingIgnoringWhitespace
+import org.jetbrains.kotlin.psi.psiUtil.getParentOfType
+import org.jetbrains.kotlin.resolve.calls.model.ResolvedCall
+import org.sonar.check.Rule
+import org.sonarsource.kotlin.api.CallAbstractCheck
+import org.sonarsource.kotlin.api.FunMatcher
+import org.sonarsource.kotlin.plugin.KotlinFileContext
+
+@Rule(key = "S6611")
+class MapValuesShouldBeAccessedSafelyCheck : CallAbstractCheck() {
+
+    override val functionsToVisit = listOf(
+            FunMatcher {
+                withQualifiers("kotlin.collections.Map", "kotlin.collections.MutableMap")
+                withNames("get", "getValue", "getOrElse", "getOrDefault", "null")
```

Comment:
I think there was a misunderstanding from the rule description.

The rule should raise an issue only on case of `map.get(key)!!` or `map[key]!!`
All other functions are considered valid as they handle nullable type. So, in this matcher you should only visit `get`. So you use `FunMatcher(name = "get")`

Also, in your matcher you're matching a direct type, however I can extend Map and have my own class, we should still raise an issue in this case, so you need to use `definingSupertypes` instead of `qualifiers`

Feel free to update the rule descrition, if you feel it's not clear enough: https://github.com/SonarSource/rspec/pull/1799/files